Hostel 13 II реализовать репозиторий для тайм слотов

Gazcoin1 opened the pull request:
Написал методы поиска с запросами query. Написал тесты для репозитория. БД для тестов запускаются в контейнерах, Для запуска необходим docker.

Gazcoin1 committed:
feat(HOSTEL-13): Реализован репозиторий слотов времени.

- Реализован интерфейс TimeSlotRepository.

Gazcoin1 committed:
feat(HOSTEL-13): Написаны тесты для TimeSlotRepository

- Написаны тесты, запускаемые в контейнере, для методов из интерфейса TimeSlotRepository.

TpaBKa251 commented on src/main/java/ru/tpu/hostel/schedules/repository/TimeSlotRepository.java:
```diff
@@ -16,16 +16,16 @@
 public interface TimeSlotRepository extends JpaRepository<TimeSlot, UUID> {
 
-    @Query
+    @Query("SELECT t FROM TimeSlot t WHERE t.type = :eventType ORDER BY t.startTime DESC LIMIT 1")
```
Не нравится мне такое написание sql-запросов. Мб сделать класс с константами пакетной видимости, где будут все эти запросы? Либо у великого чатгпт спросить, как лучше это сделать

TpaBKa251 commented on src/test/java/ru/tpu/hostel/schedules/repository/TimeSlotRepositoryTest.java:
```diff
@@ -0,0 +1,151 @@
+
+        Optional<TimeSlot> timeSlot = timeSlotRepository.findLastByType(EventType.GYM);
+
+        assertTrue(timeSlot.isPresent());
```
Можно просто вызвать assertEquals() на timeSlot. Но возможно предварительно нужно реализовать там equals() и hashCode(). Соответствующие аннотации есть у Ломбока, либо Data поставить

TpaBKa251 replied:
```
TimeSlot expected = getNewTimeSlot();
TimeSlot actual = timeSlotRepository.findLastByType().orElseThrow();

assertThat(actual)
    .usingRecursiveComparison()
    .ignoringFields("id")
    .isEqualTo(expected);
```

TpaBKa251 commented on src/test/java/ru/tpu/hostel/schedules/repository/TimeSlotRepositoryTest.java:
```diff
@@ -0,0 +1,151 @@
+
+        List<TimeSlot> timeSlots = timeSlotRepository.findAllByTypeAndStartTimeAfter(EventType.GYM, startTime);
+
+        assertEquals(1, timeSlots.size());
```
Может здесь лучше подойдёт assertThat(List).containsExactlyElementsOf(елементы)? И вообще везде, где дело с коллекциями имеем использовать подобные методы (в этом же классе выше и ниже). Но точно как там метод называется не помню🙃

TpaBKa251 replied:
```
assertThat(actualList)
                .usingElementComparatorIgnoringFields("id")
                .containsExactlyElementsOf(expectedList);
```

TpaBKa251 commented on src/test/java/ru/tpu/hostel/schedules/repository/TimeSlotRepositoryTest.java:
```diff
@@ -0,0 +1,151 @@
+        timeSlotRepository.deleteAll();
+
+        timeSlotRepository.save(getNewTimeSlot(
+                LocalDateTime.of(2025, 3, 10, 9, 0),
```
Давай заведем класс Data, куда будем пихать всякие константы, утилитарные методы и тд, необходимые для тестов. Ну и сразу туда перенести здешние даты, мб метод для генерации слотов.

Gazcoin1 committed:
feat(HOSTEL-13): Внесены изменения

- Создан класс Data с константами и статическими методами для использования их в тестах.
- Внесены правки в вид sql запросов в аннотации @Query
- Вместо @Getter и @Setter в entity стоит аннотация @Data

TpaBKa251 commented on src/test/java/ru/tpu/hostel/schedules/Data.java:
```diff
@@ -0,0 +1,37 @@
+import java.time.LocalDateTime;
+
+public class Data {
+    public static final int YEAR_2020 = 2020;
```
Можно тут сразу константы LocalDateTime добавить, чтобы тесты менее громоздко выглядели

Gazcoin1 committed:
fix(HOSTEL-13): Внесены изменения

- LocalDateTime объекты вынесены в константы.
- Переделаны тесты для более корректной проверки.

Gazcoin1 committed:
fix(HOSTEL-13): Поправка в тестах
